source/adapters/level_zero/v2/event.cpp

```diff
@@ -300,4 +361,32 @@ ur_result_t urEventGetProfilingInfo(
 } catch (...) {
   return exceptionToResult(std::current_exception());
 }
+
+ur_result_t urEventGetNativeHandle(ur_event_handle_t hEvent,
+                                   ur_native_handle_t *phNativeEvent) try {
+  *phNativeEvent = reinterpret_cast<ur_native_handle_t>(hEvent->getZeEvent());
+  return UR_RESULT_SUCCESS;
+} catch (...) {
+  return exceptionToResult(std::current_exception());
+}
+
+ur_result_t
+urEventCreateWithNativeHandle(ur_native_handle_t hNativeEvent,
+                              ur_context_handle_t hContext,
+                              const ur_event_native_properties_t *pProperties,
+                              ur_event_handle_t *phEvent) try {
+  if (!hNativeEvent) {
+    assert((hContext->nativeEventsPool.getFlags() & v2::EVENT_FLAGS_COUNTER) ==
+           0);
+
+    *phEvent = hContext->nativeEventsPool.allocate();
+    ZE2UR_CALL(zeEventHostSignal, ((*phEvent)->getZeEvent()));
+  } else {
+    *phEvent = new ur_native_event_t(hNativeEvent, hContext, pProperties);
+  }
+  return UR_RESULT_SUCCESS;
```

Comment:
`(*phEvent)->retain();`

Reply:
That should not be needed. I think the problem was with Mutex not being unlocked before destroying the event. Should be fixed now.

Reply:
I verified, it works.